libbeat/template/client_handler.go

```diff
@@ -0,0 +1,229 @@
+	Load(config TemplateConfig, info beat.Info, fields []byte, migration bool) error
+}
+
+// ESClientHandler implements ClientHandler interface for loading templates to Elasticsearch.
+type ESClientHandler struct {
+	client ESClient
+}
+
+// ESClient is a subset of the Elasticsearch client API capable of
+// loading the template.
+type ESClient interface {
+	Request(method, path string, pipeline string, params map[string]string, body interface{}) (int, []byte, error)
+	GetVersion() common.Version
+}
+
+// NewESClientHandler creates a new template loader for ES
+func NewESClientHandler(client ESClient) *ESClientHandler {
+	return &ESClientHandler{client: client}
+}
+
+// Load checks if the index mapping template should be loaded
+// In case the template is not already loaded or overwriting is enabled, the
+// template is built and written to index
+func (l *ESClientHandler) Load(config TemplateConfig, info beat.Info, fields []byte, migration bool) error {
+	//build template from config
+	tmpl, err := template(config, info, l.client.GetVersion(), migration)
+	if err != nil || tmpl == nil {
+		return err
+	}
+
+	// Check if template already exist or should be overwritten
+	templateName := tmpl.GetName()
+	if config.JSON.Enabled {
+		templateName = config.JSON.Name
+	}
+
+	if l.templateExists(templateName) && !config.Overwrite {
+		logp.Info("Template %s already exists and will not be overwritten.", templateName)
+		return nil
+	}
+
+	//loading template to ES
+	body, err := buildBody(tmpl, config, fields)
+	if err != nil {
+		return err
+	}
+	if err := l.loadTemplate(templateName, body); err != nil {
+		return fmt.Errorf("could not load template. Elasticsearch returned: %v. Template is: %s", err, body.StringToPrint())
+	}
+	logp.Info("template with name '%s' loaded.", templateName)
+	return nil
+}
+
+// loadTemplate loads a template into Elasticsearch overwriting the existing
+// template if it exists. If you wish to not overwrite an existing template
+// then use CheckTemplate prior to calling this method.
+func (l *ESClientHandler) loadTemplate(templateName string, template map[string]interface{}) error {
+	logp.Info("Try loading template %s to Elasticsearch", templateName)
+	path := "/_template/" + templateName
+	params := esVersionParams(l.client.GetVersion())
+	status, body, err := l.client.Request("PUT", path, "", params, template)
+	if err != nil {
+		return fmt.Errorf("couldn't load template: %v. Response body: %s", err, body)
+	}
+	if status > 300 {
+		return fmt.Errorf("couldn't load json. Status: %v", status)
+	}
+	return nil
+}
+
+// templateExists checks if a given template already exist. It returns true if
+// and only if Elasticsearch returns with HTTP status code 200.
+func (l *ESClientHandler) templateExists(templateName string) bool {
+	if l.client == nil {
+		return false
+	}
+	status, _, _ := l.client.Request("HEAD", "/_template/"+templateName, "", nil, nil)
+
+	if status != 200 {
+		return false
+	}
+
+	return true
```

Comment:
`return status == 200`. Better to not use magic numbers, but `return status == http.StatusOK`. No need for artificial newlines after call to `Request`.